### PR TITLE
 tests: Check that command_line_args round-trips.

### LINKED ISSUE #1818: Better testing for argument serialization / builder_from_args
I think tests.rs should (maybe only on automation to not duplicate the test runtime, or behind a test-only env var or something?) run bindgen twice, once with the result from `Builder::generate`, and once from the result of calling `builder_from_args(builder.command_line_flags())` , and ensure that their result is identical.

This will catch mistakes like #1816 in a much more thorough way than adding tests for each combination that happened to be broken at some point.

### PR DESCRIPTION
**Pull request by emilio**
Fixes #1818.

### CODE REVIEW

**Comment by emilio** [comment]
@kulp and/or @tmfink can you take a look given we've all recently looked at this code? :)

This makes me a bit more comfortable about `Builder::command_line_args` (and of course caught a couple bugs)

File: tests/tests.rs [hunk 1]
```diff
@@ -194,6 +196,14 @@ fn compare_generated_header(
         ),
     };
 
+    let flags = if check_roundtrip {
```
Comment (tmfink) [inline tests/tests.rs:199]: It looks like `flags` is not used if we are not doing the roundtrip test.
Could you move the definition under `if check_roundtrip { ...`?

File: tests/tests.rs [hunk 2]
```diff
@@ -328,11 +349,7 @@ macro_rules! test_header {
         fn $function() {
             let header = PathBuf::from($header);
             let result = create_bindgen_builder(&header).and_then(|builder| {
-                if let Some(builder) = builder {
-                    compare_generated_header(&header, builder)
-                } else {
-                    Ok(())
-                }
+                compare_generated_header(&header, builder, true)
```
Comment (tmfink) [inline tests/tests.rs:352]: How much extra time does the roundtrip add to testing? We may want to make this based on an environment variable like `BINDGEN_TEST_ROUNDTRIP`.

We could:
1. Make sure to set the env variable in CI tests
2. Adding extra logging that makes it clear we are doing roundtrip tests

File: tests/tests.rs [hunk 3]
```diff
@@ -207,37 +217,48 @@ fn compare_generated_header(
...
+
+        return Err(Error::new(ErrorKind::Other, "Header and binding differ! Run with BINDGEN_OVERWRITE_EXPECTED=1 in the environment to automatically overwrite the expectation."));
     }
 
-    // Overwrite the expectation with actual output.
-    if env::var_os(OVERWRITE_ENV_VAR).is_some() {
-        let mut expectation_file = fs::File::create(looked_at.last().unwrap())?;
-        expectation_file.write_all(actual.as_bytes())?;
+    if check_roundtrip {
+        let roundtrip_builder = builder_from_flags(flags.into_iter())?.0;
```
Comment (tmfink) [inline tests/tests.rs:253]: FYI, there are cases the bindgen program would given you an error but `builder_from_flags()` will not show an error.
Sadly, to properly handle, I think we need to execute a sub-process.

https://github.com/rust-lang/rust-bindgen/pull/1819#discussion_r446752135

File: tests/tests.rs (hunk 1 above)
Comment (emilio) [reply tests/tests.rs:199]: I can't, because `builder` is moved below in the `generate()` call.

File: tests/tests.rs (hunk 2 above)
Comment (emilio) [reply tests/tests.rs:352]: Yeah, I thought of this and I was a bit torn... On one hand, I'd like to make running the right set of tests if you're a contributor just `cargo test`. On the other I agree that this is unlikely to catch a lot of errors during development.

For me `cargo test` is still fast with this change, but I have a very fast machine... I think I'd prefer for now not moving it behind an env variable, just under the principle that "if `cargo test` is green, CI should be green too". But I'm happy to reconsider if this regresses stuff on slower machines significantly.

File: tests/tests.rs (hunk 3 above)
Comment (emilio) [reply tests/tests.rs:253]: Right, this _ought_ to work :)

File: tests/tests.rs (hunk 2 above)
Comment (kulp) [reply tests/tests.rs:352]: I have a very slow machine, but even I would rather have `cargo test` match CI by default, so I personally am fine with this the way it is.

On the other hand, if I learned how to profile (#939), then maybe I would see how much of my life I was losing.

Comment (tmfink) [reply tests/tests.rs:352]: We could also make the an opt-out environment variable like `NO_BINDGEN_TEST_ROUNDTRIP`. That way, developers who want to iterate more quickly can disable this test.

Comment (emilio) [reply tests/tests.rs:352]: Yeah, that sounds better. I'll go for `BINDGEN_DISABLE_ROUNDTRIP_TEST` or something of that sort if you don't mind, to keep the `BINDGEN_` prefix.

Comment (tmfink) [reply tests/tests.rs:352]: Sure--sounds good.

**Pull request merged**